elastic4s-http/src/main/scala/com/sksamuel/elastic4s/http/ResponseHandler.scala

```diff
@@ -0,0 +1,67 @@
+package com.sksamuel.elastic4s.http
+
+import java.nio.charset.Charset
+
+import com.fasterxml.jackson.annotation.JsonInclude
+import com.fasterxml.jackson.core.JsonParser
+import com.fasterxml.jackson.databind.{DeserializationFeature, ObjectMapper}
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+import com.fasterxml.jackson.module.scala.experimental.ScalaObjectMapper
+
+object JacksonSupport {
+
+  val mapper: ObjectMapper with ScalaObjectMapper = new ObjectMapper with ScalaObjectMapper
+  mapper.registerModule(DefaultScalaModule)
+
+  mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL)
+  mapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+  mapper.configure(DeserializationFeature.FAIL_ON_IGNORED_PROPERTIES, false)
+  mapper.configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY, true)
+  mapper.configure(JsonParser.Feature.ALLOW_UNQUOTED_FIELD_NAMES, true)
+  mapper.configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true)
+}
+
+import org.apache.http.HttpEntity
+import org.elasticsearch.client.{Response, ResponseException}
+
+import scala.io.{Codec, Source}
+import scala.util.{Failure, Try}
+
+trait ResponseHandler[U] {
+  def onResponse(response: Response): Try[U]
+  def onError(e: Exception): Try[U]
+}
+
+// a ResponseHandler that marshalls the body into the required type using Jackson
+// the response body is converted into a string using a codec derived from the content encoding header
+// if the content encoding header is null, then UTF-8 is assumed
+object ResponseHandler {
+
+  def fromEntity[U: Manifest](entity: HttpEntity): Try[U] = {
+    Try {
+      val charset = Option(entity.getContentEncoding).map(_.getValue).getOrElse("UTF-8")
+      implicit val codec = Codec(Charset.forName(charset))
+      val body = Source.fromInputStream(entity.getContent).mkString
+      JacksonSupport.mapper.readValue[U](body)
+    }
+  }
+
+  def default[U: Manifest] = new DefaultResponseHandler[U]
+  def failure404[U: Manifest] = new NotFound404ResponseHandler[U]
+}
+
+class DefaultResponseHandler[U: Manifest] extends ResponseHandler[U] {
+  override def onResponse(response: Response): Try[U] = ResponseHandler.fromEntity[U](response.getEntity)
+  override def onError(e: Exception): Try[U] = Failure(e)
+}
+
+class NotFound404ResponseHandler[U: Manifest] extends DefaultResponseHandler[U] {
```

Comment:
If you extend the `DefaultResponseHandler` then `onResponse` can be removed and you should call `super.onResponse` for the "default" case